src/query/api/v1/handler/placement/common.go

```diff
@@ -134,14 +292,120 @@ func ConvertInstancesProto(instancesProto []*placementpb.Instance) ([]placement.
+
+	// Add
+	addFn := applyMiddleware(NewAddHandler(opts).ServeHTTP)
+	r.HandleFunc(DeprecatedM3DBAddURL, addFn).Methods(AddHTTPMethod)
+	r.HandleFunc(M3DBAddURL, addFn).Methods(AddHTTPMethod)
+	r.HandleFunc(M3AggAddURL, addFn).Methods(AddHTTPMethod)
+
+	// Delete
+	deleteFn := applyMiddleware(NewDeleteHandler(opts).ServeHTTP)
+	r.HandleFunc(DeprecatedM3DBDeleteURL, deleteFn).Methods(DeleteHTTPMethod)
+	r.HandleFunc(M3DBDeleteURL, deleteFn).Methods(DeleteHTTPMethod)
+	r.HandleFunc(M3AggDeleteURL, deleteFn).Methods(DeleteHTTPMethod)
+}
+
+// We want to generate a function that will return the cutoff such that it is always at the beginning
+// of an aggregation window size, but also later than or equal to the current time + warmup. We accomplish
+// this by adding the warmup time and the max aggregation window size to the current time, and then truncating
+// to the max aggregation window size. This ensure that we always return a time that is at the beginning of an
+// aggregation window size, but is also later than now.Add(warmup).
+func newShardCutOffNanosFn(now time.Time, maxAggregationWindowSize, warmup time.Duration) placement.TimeNanosFn {
+	return func() int64 {
+		return now.
+			Add(warmup).
+			Add(maxAggregationWindowSize).
+			Truncate(maxAggregationWindowSize).
+			Unix()
+	}
+}
+
+func newShardCutOverValidationFn(now time.Time) placement.ShardValidationFn {
+	return func(s shard.Shard) error {
+		switch s.State() {
+		case shard.Initializing:
+			if s.CutoverNanos() > now.UnixNano() {
+				return fmt.Errorf("could not mark shard %d available before cutover time %v", s.ID(), time.Unix(0, s.CutoverNanos()))
+			}
+			return nil
+		default:
+			return fmt.Errorf("could not mark shard %d available, invalid state %s", s.ID(), s.State().String())
+		}
+	}
+}
+
+func newShardCutOffValidationFn(now time.Time, maxAggregationWindowSize time.Duration) placement.ShardValidationFn {
+	return func(s shard.Shard) error {
+		switch s.State() {
+		case shard.Leaving:
+			if s.CutoffNanos() > now.UnixNano()-maxAggregationWindowSize.Nanoseconds() {
+				return fmt.Errorf("could not return leaving shard %d with cutoff time %v, max aggregation window %v",
+					s.ID(), time.Unix(0, s.CutoffNanos()), maxAggregationWindowSize)
+			}
+			return nil
+		default:
+			return fmt.Errorf("could not mark shard %d available, invalid state %s", s.ID(), s.State().String())
+		}
+	}
+}
 
-	r.HandleFunc(InitURL, logged(NewInitHandler(client, cfg)).ServeHTTP).Methods(InitHTTPMethod)
-	r.HandleFunc(GetURL, logged(NewGetHandler(client, cfg)).ServeHTTP).Methods(GetHTTPMethod)
-	r.HandleFunc(DeleteAllURL, logged(NewDeleteAllHandler(client, cfg)).ServeHTTP).Methods(DeleteAllHTTPMethod)
-	r.HandleFunc(AddURL, logged(NewAddHandler(client, cfg)).ServeHTTP).Methods(AddHTTPMethod)
-	r.HandleFunc(DeleteURL, logged(NewDeleteHandler(client, cfg)).ServeHTTP).Methods(DeleteHTTPMethod)
+type placementCutoverOpts struct {
```

Comment:
nit: rename to `m3aggregatorPlacementOpts`?

Reply:
done